### DISCUSSION TITLE
SQL Alchemy Enum + native_enum disabled : not raising the expected exception

### DISCUSSION THREAD

**Discussion by jossefaz** [sqlalchemy/sqlalchemy#7181] [Usage Questions] [answered]
For all bellow examples I used 

sqlalchemy 1.4

python 3.9

MySQL 5.7.17

I define an engine and a session object :

```python
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

engine = create_engine("mysql+pymysql://test:test@localhost:3306/test")
Session = sessionmaker(bind=engine)
```

I then defined a Base table that contains an Enum type with the parameter [`native_enum`][1] set to `False` (on purpose because I do not want an ENUM type at database level but **just want from sqlalchemy ORM to act as a data gatekeeper**) :

```python
from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy import Enum as SaEnum
from sqlalchemy.ext.declarative import declarative_base
import enum

Base = declarative_base()

class DummyEnum(enum.Enum):
    NAME_1 = "NAME_1"
    NAME_2 = "NAME_2"

class SAEnumVarcharTable(Base):
    __tablename__ = "test_table"
    id = Column(Integer, primary_key=True)
    uut_column = Column(SaEnum(DummyEnum, native_enum=False), default=DummyEnum.NAME_1)
```

I then try to insert a row with a value that is not defined in `DummyEnum`.

```python
with Session() as session:
    dt = SAEnumVarcharTable(uut_column="NAME_5")
    session.add(dt)
    session.commit()
```
And it bypasses the expected control from the SQL Alchemy enum type.
No exception is raised and the row is inserted in the database.

I expected SqlAlchemy to raise any exception in this case...

Doc on SQL Alchemy Enum type : [here][2]

  [1]: https://docs.sqlalchemy.org/en/14/core/type_basics.html#sqlalchemy.types.Enum.params.native_enum
  [2]: https://docs.sqlalchemy.org/en/14/core/type_basics.html#sqlalchemy.types.Enum

**Comment by CaselIT** [comment, marked as answer]
Hi,

As the documentation mentions, both [`create_constraint`](https://docs.sqlalchemy.org/en/14/core/type_basics.html#sqlalchemy.types.Enum.params.create_constraint) and [`validate_strings`](https://docs.sqlalchemy.org/en/14/core/type_basics.html#sqlalchemy.types.Enum.params.validate_strings) are false by default, so no validation happens if the db does not support enums.

Enabling one and/or the other should cause the appropriate error to be raised

**Reply by jossefaz** [reply]
Thank you very much for your answer.
The MySQL version that I use supports Enums.
And I tested it with the same piece of code as above...but without the `native_enum` arg set to False.
And the validation occurs, it created an ENUM column in the DB and does not allow other values to be inserted.
This is why by just adding this param I thought that it will still act as a gatekeeper.

So if I understand correctly, adding `native_enums` will transform my column as a plain VARCHAR column ? I don't get what is the purpose of this parameter then...why would I create a table with an enum type column if it does not do any validation ?

I tried what you suggested. It works well. But adding `validate_string` only was sufficient to raise the error : 

```python 
 uut_column = Column(SaEnum(DummyEnum, native_enum=False, validate_strings=True), default=DummyEnum.NAME_1)
```

Should I add `create_constraint` as well ? I mean what would that add ?
From the docs I understand that `validate_strings` is for code validation and `create_constraints` is for ensuring a DB validation as well...Am I correct ?

**Reply by CaselIT** [reply]
> So if I understand correctly, adding native_enums will transform my column as a plain VARCHAR column ? I don't get what is the purpose of this parameter then...why would I create a table with an enum type column if it does not do any validation ?

Some db do not support enum natively. Until 1.4 the default was to have `create_constraint=True` but it usually causes issue with migrations and general updates to the db, so it was set to `False` by default. Here is the change https://docs.sqlalchemy.org/en/14/changelog/changelog_14.html#change-22b21b44239fae0edb1b03961bb883d4

Regarding what you should use, it depends on your use case. `validate_strings` is done in python, so when accessing the db directly you can insert different strings, while `create_constraint` generates check constraint on the db side.

I think it usually makes sense having one or the other, both seem unnecessary.

Note that `create_constraint` acts only when creating the table, so it will not apply any change to an existing table

**Reply by jossefaz** [reply]
By the way : the `create_constraint` param will not create the CHECK contraint for MySQL 5.1.17. In MySQL 8.0.26 it did create the contraint
Thank you very much for your help

**Reply by CaselIT** [reply]
It seems that mysql before 8.0.16 did not support check constraint, so that would be the reason I think.

I'm not really a mysql expert
